File: ios/VideoComposer.m
#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>
#import <AVKit/AVKit.h>

@interface VideoComposer : RCTEventEmitter <RCTBridgeModule>
{
  NSMutableDictionary *_compositionsData;
}
@end

@implementation VideoComposer

RCT_EXPORT_MODULE();

static RCTEventEmitter* staticEventEmitter = nil;

+ (BOOL)requiresMainQueueSetup {
    return NO;
}

- (id) init {
  self = [super init];
  if (self) {
    staticEventEmitter = self;
    
    _compositionsData = [NSMutableDictionary dictionary];
  }
  return self;
}


- (void)_sendEventWithName:(NSString *)eventName body:(id)body {
  if (staticEventEmitter == nil)
    return;

  [staticEventEmitter sendEventWithName:eventName body:body];
}

- (NSArray<NSString *> *)supportedEvents {
    return @[
        @"@clipisode/react-native-video-composer:progress",
        @"@clipisode/react-native-video-composer:error"
    ];
}

// --------------------
// -- IMPLEMENTATION --
// --------------------

RCT_EXPORT_METHOD(compose:(NSDictionary *)composition id:(NSString *)exportId outPath:(NSString *)outPath resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
  // SETUP DATA OBJECT AND CHECK FOR EXISTING EXPORT
  
  NSMutableDictionary *compositionData = _compositionsData[exportId];
  
  if (compositionData) {
    reject(@"Duplicate", @"A composition with this exportId is already in progress.", nil);
    return;
  }
  
  compositionData = [NSMutableDictionary dictionary];
  [_compositionsData setObject:compositionData forKey:exportId];
  
  // BEGIN COMPOSITION SETUP
  
  AVMutableComposition *mixComposition = [[AVMutableComposition alloc] init];
  AVMutableCompositionTrack *videoTrack = [mixComposition addMutableTrackWithMediaType:AVMediaTypeVideo preferredTrackID:kCMPersistentTrackID_Invalid];
  
  CMTime lastEndTime = kCMTimeZero;
  
  NSArray *videos = composition[@"videos"];
  
  for (NSDictionary* video in videos) {
    NSString *path = video[@"path"];
    // NSNumber* startAt = video[@"startAt"];
    
    NSURL *url = [NSURL URLWithString:path];
    AVAsset *asset = [AVAsset assetWithURL:url];
    
    CMTimeRange range = CMTimeRangeMake(kCMTimeZero, asset.duration);
    NSArray* assetVideoTracks = [asset tracksWithMediaType:AVMediaTypeVideo];
    AVAssetTrack* firstVideoTrack = assetVideoTracks[0];
    [videoTrack insertTimeRange:range ofTrack:firstVideoTrack atTime:lastEndTime error:nil];
          
    lastEndTime = CMTimeAdd(lastEndTime, asset.duration);
  }
  
  AVMutableVideoCompositionInstruction *mainInstruction = [[AVMutableVideoCompositionInstruction alloc] init];
  mainInstruction.timeRange = CMTimeRangeMake(kCMTimeZero, lastEndTime);
  
  AVMutableVideoCompositionLayerInstruction *videolayerInstruction = [AVMutableVideoCompositionLayerInstruction videoCompositionLayerInstructionWithAssetTrack:videoTrack];
  [videolayerInstruction setOpacity:0.0 atTime:lastEndTime];
  
  mainInstruction.layerInstructions = [NSArray arrayWithObjects:videolayerInstruction, nil];

  AVMutableVideoComposition *mainCompositionInst = [AVMutableVideoComposition videoComposition];
  
  mainCompositionInst.renderSize = CGSizeMake(720.0, 1280.0);
  
  mainCompositionInst.instructions = [NSArray arrayWithObject:mainInstruction];
  mainCompositionInst.frameDuration = CMTimeMake(1, 30);
  
  
  // -- EXPORT
  
  AVAssetExportSession *exporter = [AVAssetExportSession exportSessionWithAsset:mixComposition presetName:AVAssetExportPresetHighestQuality];
  [compositionData setObject:exporter forKey:@"exportSession"];
  
  exporter.outputURL = [NSURL URLWithString:outPath];
  exporter.outputFileType = AVFileTypeMPEG4;
  exporter.shouldOptimizeForNetworkUse = YES;
  exporter.videoComposition = mainCompositionInst;
  [exporter exportAsynchronouslyWithCompletionHandler:^{
    dispatch_async(dispatch_get_main_queue(), ^{
      resolve(nil);
    });
  }];
  
  dispatch_async(dispatch_get_main_queue(), ^{
    NSTimer *timer = [NSTimer scheduledTimerWithTimeInterval:0.1 repeats:YES block:^(NSTimer * _Nonnull timer) {
      if (exporter.status == AVAssetExportSessionStatusExporting) {
        NSMutableDictionary *body = [NSMutableDictionary dictionary];
        [body setObject:exportId forKey:@"id"];
        [body setObject:[NSNumber numberWithFloat:exporter.progress] forKey:@"progress"];
          
        [self sendEventWithName:@"@clipisode/react-native-video-composer:progress" body:body];
      } else if (exporter.status == AVAssetExportSessionStatusCompleted) {
        [timer invalidate];
        [self->_compositionsData removeObjectForKey:exportId];
      }
    }];
    [compositionData setObject:timer forKey:@"timer"];
  });
}

RCT_EXPORT_METHOD(cancel:(NSString *)exportId)
{
  NSMutableDictionary *compositionData = _compositionsData[exportId];
  
  if (!compositionData) {
    NSLog(@"Composition data not found for given exportId.");
  } else {
    AVAssetExportSession *exportSession = compositionData[@"exportSession"];
    NSTimer *timer = compositionData[@"timer"];
    
    if (timer) {
      [timer invalidate];
    }
    
    if (exportSession) {
      [exportSession cancelExport];
    }
    
    [_compositionsData removeObjectForKey:exportId];
  }
}

@end

File: ios/VideoComposer.swift
import AVKit

@objc(VideoComposer)
class VideoComposer: NSObject {
  override init() {
  }
  
  @objc
  func compose(_ composition: NSDictionary, outPath outputPath: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) -> Void {
    let videos: NSArray? = composition["videos"] as? NSArray;
    
    if (videos != nil && videos!.count > 0) {
      build(videos: videos!, outputPath: outputPath, resolver: resolve)
    } else {
      resolve(nil);
    }
  }
  
  func build(videos: NSArray, outputPath: String, resolver resolve: @escaping RCTPromiseResolveBlock) {
    let mixComposition = AVMutableComposition();
    
    let videoTrack = mixComposition.addMutableTrack(withMediaType: AVMediaType.video, preferredTrackID: kCMPersistentTrackID_Invalid);
    
    var lastEndTime: CMTime = .zero;
    
    for video in videos {
      let path = (video as! NSDictionary)["path"] as! String;
      let startAt = (video as! NSDictionary)["startAt"] as! NSNumber;
      
      let url = URL.init(string: path)!;
      let asset = AVAsset.init(url: url);
        
      let range: CMTimeRange = CMTimeRangeMake(start: .zero, duration: asset.duration);
      
      do {
        let assetVideoTracks: [AVAssetTrack] = asset.tracks(withMediaType: .video)
        let firstVideoTrack: AVAssetTrack = assetVideoTracks.first!;
        
        try videoTrack!.insertTimeRange(range, of: firstVideoTrack, at: lastEndTime)
      } catch { print("test"); };
            
      lastEndTime = CMTimeAdd(lastEndTime, asset.duration);
    }
  
    // 3.1 - Create AVMutableVideoCompositionInstruction
    let mainInstruction = AVMutableVideoCompositionInstruction()
    mainInstruction.timeRange = CMTimeRangeMake(start: .zero, duration: lastEndTime);

    // 3.2 - Create an AVMutableVideoCompositionLayerInstruction for the video track and fix the orientation.
    let videolayerInstruction = AVMutableVideoCompositionLayerInstruction(assetTrack: videoTrack!)
//    let videoAssetTrack = asset.tracks(withMediaType: .video)[0]
//    var videoAssetOrientation_: UIImage.Orientation = .up
//    var isVideoAssetPortrait_ = false
//    let videoTransform = videoAssetTrack.preferredTransform
//    if videoTransform.a == 0 && videoTransform.b == 1.0 && videoTransform.c == -1.0 && videoTransform.d == 0 {
//        videoAssetOrientation_ = UIImage.Orientation.right
//        isVideoAssetPortrait_ = true
//    }
//    if videoTransform.a == 0 && videoTransform.b == -1.0 && videoTransform.c == 1.0 && videoTransform.d == 0 {
//        videoAssetOrientation_ = UIImage.Orientation.left
//        isVideoAssetPortrait_ = true
//    }
//    if videoTransform.a == 1.0 && videoTransform.b == 0 && videoTransform.c == 0 && videoTransform.d == 1.0 {
//        videoAssetOrientation_ = UIImage.Orientation.up
//    }
//    if videoTransform.a == -1.0 && videoTransform.b == 0 && videoTransform.c == 0 && videoTransform.d == -1.0 {
//        videoAssetOrientation_ = UIImage.Orientation.down
//    }
//    videolayerInstruction.setTransform(videoAssetTrack.preferredTransform, at: .zero)
    videolayerInstruction.setOpacity(0.0, at: lastEndTime)

    // 3.3 - Add instructions
    mainInstruction.layerInstructions = [videolayerInstruction]
    


    let mainCompositionInst = AVMutableVideoComposition()

//    var naturalSize: CGSize
//    if isVideoAssetPortrait_ {
//        naturalSize = CGSize(width: videoAssetTrack.naturalSize.height, height: videoAssetTrack.naturalSize.width)
//    } else {
//        naturalSize = videoAssetTrack.naturalSize
//    }
//
//
//    var renderWidth: Float
//    var renderHeight: Float
//    renderWidth = Float(naturalSize.width)
//    renderHeight = Float(naturalSize.height)
    mainCompositionInst.renderSize = CGSize(width: CGFloat(720), height: CGFloat(1280))
    mainCompositionInst.instructions = [mainInstruction]
    mainCompositionInst.frameDuration = CMTimeMake(value: 1, timescale: 30)

    // 5 - Create exporter
    let exporter = AVAssetExportSession(asset: mixComposition, presetName: AVAssetExportPresetHighestQuality)
    exporter?.outputURL = URL(string: outputPath)
    exporter?.outputFileType = .mp4
    exporter?.shouldOptimizeForNetworkUse = true
    exporter?.videoComposition = mainCompositionInst
    exporter?.exportAsynchronously(completionHandler: {
        DispatchQueue.main.async(execute: {
          resolve(nil);
        })
    })
  }
}
